src/tests/unit_asio_stream.cpp

```diff
@@ -0,0 +1,772 @@
+         {
+         net::io_context ioc;
+         // fail right away
+         FailCount  fc{0, net::error::eof};
+         TestStream remote{ioc};
+
+         Botan::TLS::Context ctx;
+         AsioStream ssl(ctx, ioc, fc);
+         ssl.next_layer().connect(remote);
+
+         Test::Result result("async write_some error");
+
+         auto write_handler = [&](const error_code &ec, std::size_t bytes_transferred)
+            {
+            result.test_eq("committed some bytes to the core", bytes_transferred, TEST_DATA_SIZE);
+            result.confirm("propagates error code", ec == net::error::eof);
+            };
+
+         net::async_write(ssl, net::const_buffer(TEST_DATA, TEST_DATA_SIZE), write_handler);
+
+         ioc.run();
+         results.push_back(result);
+         }
+
+      void test_async_write_throw(std::vector<Test::Result>& results)
+         {
+         net::io_context ioc;
+         TestStream remote{ioc};
+
+         Botan::TLS::Context ctx;
+         ThrowingAsioStream ssl(ctx, ioc);
+         ssl.next_layer().connect(remote);
+
+         Test::Result result("async write_some throw");
+
+         auto write_handler = [&](const error_code &ec, std::size_t bytes_transferred)
+            {
+            result.test_eq("didn't transfer anything", bytes_transferred, 0);
+            result.confirm("propagates error code", ec == ThrowingMockChannel::expected_ec());
+            };
+
+         net::async_write(ssl, net::const_buffer(TEST_DATA, TEST_DATA_SIZE), write_handler);
+
+         ioc.run();
+         results.push_back(result);
+         }
+
+   public:
+      std::vector<Test::Result> run() override
+         {
+         std::vector<Test::Result> results;
+
+         test_sync_handshake(results);
+         test_sync_handshake_error(results);
+         test_sync_handshake_throw(results);
+
+         test_async_handshake(results);
+         test_async_handshake_error(results);
+         test_async_handshake_throw(results);
+
+         test_sync_read_some_success(results);
+         test_sync_read_some_buffer_sequence(results);
+         test_sync_read_some_error(results);
+         test_sync_read_some_throw(results);
+         test_sync_read_zero_buffer(results);
+
+         test_async_read_some_success(results);
+         test_async_read_some_buffer_sequence(results);
+         test_async_read_some_error(results);
+         test_async_read_some_throw(results);
+         test_async_read_zero_buffer(results);
+
+         test_sync_write_some_success(results);
+         test_sync_write_some_buffer_sequence(results);
+         test_sync_write_some_error(results);
+         test_sync_write_some_throw(results);
+
+         test_async_write_some_success(results);
+         test_async_write_some_buffer_sequence(results);
+         test_async_write_some_error(results);
+         test_async_write_throw(results);
+
+         return results;
+         }
+   };
+
+BOTAN_REGISTER_TEST("asio_stream", Asio_Stream_Tests);
```

Comment:
Test should be named "tls_asio_stream" so it gets grouped with the rest of the TLS tests.

(I need to add a proper grouping mechanism to the tests but for now common prefixes suffice well enough)